A Textbox with text wrapping and box resizing (subclasses IText)

jafferhaider opened the pull request:
For [Issue 187](https://github.com/kangax/fabric.js/issues/187). This is a more fully featured implementation than [Pull Request 725](https://github.com/kangax/fabric.js/pull/725). 

I need this feature for my product, so I fully intend to support this class if it is made official.

### Demo
Check it out [a demo](http://jsfiddle.net/jafferhaider/51u08n39/92/).

### Summary of features
- Subclasses IText, so gets all of its goodness.
- Text wraps as user types and/or resizes a Textbox. Works with all textAlign values.
- User can change width of Textbox. Height is automatically set based on text wrapping. Font size and scale does not change.
- Support for resizing in Groups. Only the width of a Textbox changes. The Textbox updates its scaleX/Y values to undo scale values of its parent Group so that font size remains the same.
- Refactored the ton of `this.text.split(this._reNewline)` calls into a function that gets overwritten by subclasses.
- Fixed bug in Justify text (was present in both Text and IText classes).
- Tested on latest versions of Safari (OSX & iOS), Chrome (OSX & Win), Firefox and Internet Explorer.

### Open issues / TODOs
1. Need to add support for Styles. It is not required for my usage of this widget so contributions are welcome.
2. ~~Justify alignment needs to be fixed in IText.~~
3. ~~Inheritance bug in IText.toSVG(). The `callSuper` in `_setSVGTextLineText` goes into a recursive loop if called on a Textbox object. It works fine if I replace the `callSuper` call with a call to a copy of `Text._setSVGTextLineText`. Have you encountered something like this before? Any opinions on how to fix this?~~

Please list changes that you'd like me to make before this can be made official. I saw in the other pull request that @kangax wasn't too sure about the 'Textbox' name ... we can change it.

maximus9999 commented:
@jafferhaider and @kangax Any idea when the textbox class will be added to Fabric?  I would love to start using this functionality.  Thanks again for hard work!

kangax commented:
@maximus9999 right after 1.5 release (we'll need new or updated PR though)

jafferhaider commented:
Hey @kangax, could you please have a look at the PR now? I tried to fix the merge commit and have also updated my branch with the latest code in upstream/master.

kangax commented:
You can actually see it yourself in https://github.com/kangax/fabric.js/pull/1778/files — notice unnecessary files like HEADER.js

lookitsatravis commented:
I just want to say I'm on the edge of my seat waiting for this PR to come through. Awesome work!

jafferhaider commented:
Thanks. I've reverted changes. The following files are no longer changed in this PR:
HEADER.js
README.md
bower.json
package.json

Please let me know if I missed something @kangax.

moarsel commented:
This PR looks awesome. It works really well, except I noticed a small problem where text padding wasn't kept when saving and loading from JSON. I will see if I can replicate it in a test case in a couple days. Great work!

jafferhaider commented:
@kangax any idea what the Travis build error is about?

asturur commented:
we reverted some changes to make travis work again. it was not your code fault.

jafferhaider commented:
Thanks for the clarification @asturur. Merged latest into PR ... it looks good.

lookitsatravis commented:
Any movement on this? Still looking forward to this feature!

inssein commented:
Looks like they are waiting on https://github.com/kangax/fabric.js/pull/1872 to go through. Looking forward to this as well.

KJTsanaktsidis commented:
@jafferhaider I am using this branch in my project, but ran into an issue when enabling aspect-ratio preserving scaling on textboxes. The textbox adjusts its width but not its height, leading to very squished text due to the ScaleY. I think the width-adjustment behaviour should be enabled only during pure-X scale. I have implemented a fix for this in my fork: https://github.com/StileEducation/fabric.js/commit/6c931555272a1b1dde0b97aadfc5414b242f5b80

kusmierz commented:
looks gorgeous. awesome work. Waiting for version >1.5.0!

inssein commented:
@jafferhaider what's left before this can be merged? Looking forward to this feature and willing to help where necessary.

Noumankhan80 commented:
@jafferhaider: Asslam o ALikum,  I need your help in text wrapping can you message me I need to know what changes I should make to make itext behavior text wrapping instead of scalling and goes large. My email address is nouman_kh8@hotmail.com. Can you help me. I am on deadline of my project.

marcospassos commented:
What is missing to get it merged?

marcospassos commented:
@jafferhaider Could you please rebase with the latest version of fabric.js and update the demo? I really would like to use your PR but the version shared through fiddle is outdated and the SVG serialization is not working

KJTsanaktsidis commented:
@marcospassos what isn't working? I played with svg serialisation on this branch recently and it seemed ok to me?

marcospassos commented:
@KJTsanaktsidis Could you please share your build? I'm getting:

```
Uncaught TypeError: undefined is not a function
```
fabric.js:8016
```
fabric.createSVGFontFacesMarkup(this.getObjects()),
```

KJTsanaktsidis commented:
Hum, could you post a fiddle? I'll take a look at it later this morning

marcospassos commented:
Here is:
http://jsfiddle.net/51u08n39/61/

There are some outdated stuffs too:
```
'CanvasRenderingContext2D.webkitImageSmoothingEnabled' is deprecated. Please use 'CanvasRenderingContext2D.imageSmoothingEnabled' instead.
```

KJTsanaktsidis commented:
@marcospassos I think you just need to build fabric including the parser module- your fiddle works fine if this is done; http://jsfiddle.net/51u08n39/62/.

`build.js` seems to suggest that the parser module is used for importing, rather than exporting, SVG, but this would appear to be a complete lie, since `static_canvas.class.js` calls functions from parser in `_setSVGHeader`. I don't think this has anything to do with this PR.

KJTsanaktsidis commented:
Now that I think about it, `createSVGFontFacesMarkup` and `createSVGRefElementsMarkup` probably shouldn't be in `parser.js`, so this is probably a bug you can report.

marcospassos commented:
@jafferhaider I've found another bug. It's supposed to both have exactly the same number of lines:
http://jsfiddle.net/51u08n39/65/

marcospassos commented:
Another really critical bug: where is the style in the following text box? Now, try to resize the box. The problem is that the style is set per line but the line changes whenever you resize the box. In that, is necessary to adjust not only the line break, but also the style indexing.

http://jsfiddle.net/51u08n39/66/

marcospassos commented:
@jafferhaider  Seems like the IText and Textbox box are slightly different. Is it expected?
http://jsfiddle.net/51u08n39/68/

marcospassos commented:
About #2049, it might be a bug introduced by the feature (I'm not sure):

> I was testing the PR #1778 and I noticed that changing the style of an active object breaks the selection box. It looks like a bug once I'm calling the `setCoords()` before.

> An workaround is reset the active object which redraws correctly the selection box.

> Fiddle:
> http://jsfiddle.net/51u08n39/67/

yinxiangbo commented:
when resizing the textbox which has some styles, the text is wrapped,but the styles have some problem.
demo:http://jsfiddle.net/loveve/axvt0btu/11/

marcospassos commented:
@yinxiangbo I've [https://github.com/kangax/fabric.js/pull/1778#issuecomment-83272264](already) reported it.

marcospassos commented:
@kangax  @asturur I would like to fix the current bugs of this PR to work smoothly on master. One of them is that object scaling must be prevented in order to resize the text box instead of scaling it. Previously, it was possible overriding the `_renderNative` method that is not available anymore. Do you have any tip of how to achieve the same result on master?

kangax commented:
@marcospassos we removed Cufon option (`renderViaCufon`) so now `renderViaNative` is just `render`, and it's the only one.

sapics commented:
@marcospassos I tested to merge this PR to master branch, some days ago.
https://github.com/sapics/fabric.js/tree/jafferhaider_textbox_merge

However, I have created bugs which are not in this PR, when I merged.
A bug in my merged branch what I have found is the textbox cursor position is not same as usual.
http://jsfiddle.net/sapics/6oxexzpx/

I failed to merge.
However, if my failed branch would be some help, it is my pleasure.

jafferhaider commented:
Hey @marcospassos ... thanks for your interest in this feature. I'm struggling to keep up with new bugs that are popping up as a lot of underlying Text and IText code is being changed. I seem to have been unlucky with the timing for this PR ... I wrote all my code before these changes. :)

I'd been sidetracked with other work, but I'm back to mostly focusing on getting this code back in a stable state. I'll keep you posted. Worst case, if we decide that it is a lot of work to get Textbox working again with the new code, we might just revert back to an old version of the code and drop the plan of merging this into FabricJS. Losing a lot of time merging changes in Text and IText code into our branch. :(

asturur commented:
We could also think of including wordwrapping and fix width and height in itext, without the need to create all the other code ( controls behaviour and other ). It would look like more coherent in my opinion ( OPINION ). textbox is something that looks strange to me in a canvas related library. i repeat OPINION i do not want to get angry all the +1ers :)

jafferhaider commented:
Inheritance results in cleaner code that's faster at runtime since its not always checking control variables etc. I believe @kangax and I already had this discussion before I started work on this PR.

When are you guys going to be done with 1.5.0 so my PR can be accepted?

asturur commented:
There is this itext bug to fixed ( justify strange behavour )
And  also some json loading callback issues.

Other than that we are done.
I should go trought the list of bugs to be sure.

jafferhaider commented:
@marcospassos I've updated the demo and this PR branch with the latest code. I've also updated the list of TODOs in the description of this PR.

@kangax @asturur your new Text/IText code actually resulted in cleaner design of the Textbox class ... needed to override less methods with Textbox specific code. Great job @asturur :+1: :)

When will the bug with Justify alignment be fixed? I don't see that bug in the 1.5.1 milestone. Should I create one?

asturur commented:
I created a bug in the rendering stack. I do not want to unwind the work on the rendering, so i'm dedicating today to fix that bug ( my free time of today really ).
The "finishing" touches to itext should be done immediately after, following the issue #2109. Should not create differences with your code.

asturur commented:
@jafferhaider i wanted to tank you for the positive feedback on my work over the code, i appreciated it  and i forgot to say. There is a lot of improvement possible in Text / iText, we will do another ```round``` in the future.

jafferhaider commented:
You're welcome @asturur. Yeah I made note of a few places in the code that can be improved. Will discuss those once Textbox is stable and merged. Looking forward to working on this stuff together.

kangax commented:
I believe justify bug is mostly fixed but @asturur should know better

jafferhaider commented:
@asturur I don't see a commit related to justify. Maybe I missed something? Lemme know when I should update my fork with latest fabricjs.

asturur commented:
I think that the justify fix which kangax refers to, is the space fix we pulled before merging.
I'll try to fix some glitches during weekend, but i think you can update anyway, i have to check mainly styles deletion and clenanig of some functions. Not big changes.

DanieleSassoli commented:
Hi, I'm making some tests with Fabric 1.5.0 and this textbox which seems excellent, I have found a couple of issue though. First when justifing the text, initially the words exceed the box limit, but when I click out of the box then the text gets justified correctly. 
The second issue I found is when rendering the text on nodejs, the results are completely different from what you see on the broswer. 
This is what I see on the broswer
![broswer](https://cloud.githubusercontent.com/assets/4007741/7388543/5649d3ba-ee66-11e4-9c03-077f99fd35dd.png)
And this is how is rendered on node
![3-c0](https://cloud.githubusercontent.com/assets/4007741/7388549/68043e88-ee66-11e4-9aab-5a10cfb35115.jpg)
I just import from the exported json, wich you can find [here](http://pastebin.com/MU7Wh8cH).
Nobody else with this issue?

alanmastro commented:
Hi,
Related to the first issue you reported, I noticed that this happens only when the text isn't aligned to the right. SO if my text is alligned left and the I justify it, the text exceeds  the box limit(usually shifted to the left), if it's aligned right and then I justify it, then it's justified correctly. 
Don't know if this is of any help...

peleteiro commented:
This is great! Thank you for all your work!

inssein commented:
@jafferhaider looks like all the text styling (IText) related are fixed. It would be awesome to finally get this merged. I am guessing all that's left is rebasing right?

jafferhaider commented:
@kangax I've fixed the remaining bugs in this PR. Please review code ... I think it is ready for merge. There are a few performance related things in the Text and IText infra code that I'd like to suggest improvements for after this is merged and done with.

I haven't reviewed support for Styles at the moment, I'm hoping someone can contribute once this is merged into master.

kangax commented:
@jafferhaider I'm seeing lots of unrelated changes from the first few files. Can you please merge properly first.

josesayago commented:
@jafferhaider @kangax Hey guys, I am willing to help with this PR. I need this feature and would be great if it could finally get merged into the master branch. However, I have limited time to read through the code and find out what do I have to do.

It would be great if you could help me understand which files should I look at while trying to implement this feature. I want to know:

- Name of the files involved in this PR which must be maintained and its folders
- Dependencies to make it work (classes, functions, others)

I tried merging the justify text fix but ended with a semi-broken library, I also noticed iText and text classes were modified, but the master branch is ahead for a lots of commits from the ones this PR changes so I don't know which one to keep.

It is hard to keep going like this because FabricJS is merging lots of improvements I need but still lacks this feature which gives me a hard time merging into the latest versions of the master branch, the last successful merging I made lacks the justify patch which once I merge breaks some of the Textbox functionality.

jafferhaider commented:
@kangax please see cleaned up PR: https://github.com/kangax/fabric.js/pull/2225 ... sorry I'm having a hard time with Git ... only used it for contributing to Fabric so far :)

kangax commented:
In case someone missed, I merged all of this from another PR. Thanks to everyone involved, again!

jafferhaider commented:
Yay! Great job everyone. I see an issue with justify alignment ... will work on a solution soon.

nokoz commented:
HELP! How to remove the padding above the image? http://jsfiddle.net/rr4qxnf4/

asturur commented:
If you explain me,
why you posted here in a closed issue, with wrong topic,
i give you solution.

nokoz commented:
This indentation appeared after connecting fabric.js with the method of the Textbox

asturur commented:
you probably updated fabricjs to a newer version where the width of object is calculated counting strokewidth togheter with width.

in your script add the following line to eliminate borders for images:

fabric.Image.prototype.strokeWidth = 0;

it will go away.

When there is an unrelated issue open a new issue.
For general help refer to google group.

nokoz commented:
Ok, thanks you!

fahadnabbasi commented:
Hi Haider,

It is a great script, I was wondering if you were able to implement the text styling feature as well? I mean if a textbox has multiple styles associated with it then the calculated width takes into account those styles as well or not. 

also, with the styles, when you resize the object it doesn't take into account the style positions properly

Here is the fiddle that describes both the issues

http://jsfiddle.net/fahadnabbasi/6t0Lswf0/

Fahad

asturur commented:
Please download latest version. 1.6.0.rc1

fahadnabbasi commented:
Can you please tell me how to download only the textbox.js file? Or the
changes are made in core Fabricjs library as well?

In either case, please tell me how to download

On Mon, Aug 17, 2015 at 9:30 PM, Andrea Bogazzi <notifications@github.com>
wrote:

> Please download latest version. 1.6.0.rc1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kangax/fabric.js/pull/1778#issuecomment-131881148>.
>



-- 
Fahad Nawaz

quordev commented:
Textbox is really good but I found it's performence issue.
When I paste long text in textbox, it freeze browser minutes.
Any thoughts on this issue?

asturur commented:
check this : https://github.com/kangax/fabric.js/issue/2328

you can try it here: http://jsfiddle.net/xmfw65qg/3/

needs to be merged.

KJTsanaktsidis commented:
@quordev the last time I used the text box functionality I ran into a number of bugs in the text wrapping algorithm that caused it to take exponential time or to not terminate at all. I'm not sure if these issues have been fixed yet; we opted to fork fabric and reimplement the text box with an actual browser textarea. We decided the default browser text input behaviours were just too hard to implement in JavaScript!

quordev commented:
@KJTsanaktsidis 
I totally agree with your opinion.
I think it is not ready for production yet.

fahadnabbasi commented:
Thanks a lot Andrea, one more question please, is there a way we can get
the lineHeight property for each line instead of one for the whole textbox?

Any guidelines would be appreciated as well

On Thu, Aug 20, 2015 at 5:41 PM, Andrea Bogazzi <notifications@github.com>
wrote:

> check this : https://github.com/kangax/fabric.js/issue/2328
>
> you can try it here: http://jsfiddle.net/xmfw65qg/3/
>
> needs to be merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kangax/fabric.js/pull/1778#issuecomment-132994162>.
>



-- 
Fahad Nawaz

asturur commented:
Textbox is  just IText with autowrap functionality and disabled scale by corners.
Lineheight is a property of whole instance.

So no, there is no way to use different lineheight per line.

asturur commented:
You can style apply special style to lines of course. ( bigger font, bold, color... )

fahadnabbasi commented:
Yes I know that we can apply different styles to characters but I wanted to
add lineHeights per line, there is something called __lineHeights and if
you remove it from clearcache function, it somehow gives you the control of
each line height but still need to figure it out.

Thanks for your help :)

On Thu, Aug 20, 2015 at 6:01 PM, Andrea Bogazzi <notifications@github.com>
wrote:

> You can style apply special style to lines of course. ( bigger font, bold,
> color... )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kangax/fabric.js/pull/1778#issuecomment-133000274>.
>



-- 
Fahad Nawaz

asturur commented:
That is a cache array, is not supposed to be manipulated by user.
it would be easier to enlarge the fontsize of '\n' character at the end of line.
It would be more "stable".

fahadnabbasi commented:
Wow, a great work around, you are awesome Andrea

wendon commented:
Hi @jafferhaider ,

Thanks for great working.
I found one issue in textbox. This happens in demo also.
http://jsfiddle.net/jafferhaider/51u08n39/92/
If you click in middle of text and hit enter several times and try to add some words, cursor position seems to incorrect.

After digging source code, I guess startPosition is incorrect when original text has '\n' values.
Any thoughts on this issue?

xtecox commented:
Hi everyone,

i think i've found a bug with Selectable:false


To reproduce the bug you have to 
double click t3 
then 
drag t4 over t3
then 
drag t4 

http://jsfiddle.net/dL86Lu7k/



It shouldnt be draggable or selectable right? 

Thanks in advance

xtecox commented:
The problem is when it becomes editable

Workaround >>

set 
editable and selectable false

fahadnabbasi commented:
There is a bug. If you add a textbox and apply the formatting on selection (not whole textbox). Now if you try to align the text, it won't align

Here is the fiddle, can anyone help me?
http://jsfiddle.net/fahadnabbasi/kLdng24p/10/

The similar issue can be replicated on Kitchensink as well

Stefanoviic commented:
Having the same bug as @wendon 

"If you click in middle of text and hit enter several times and try to add some words, cursor position seems to incorrect. After digging source code, I guess startPosition is incorrect when original text has '\n' values.Any thoughts on this issue?"

Any update on this?

asturur commented:
http://jsfiddle.net/51u08n39/142/

i updated the fiddle with current fabricjs
Only cursor bug is under justify alignment with multiple spaces.

Stefanoviic commented:
@asturur 

Thanks a lot! Seems to work fine for me now.
Is there any changelog, where I can see what has changed between 1.5.0 and 1.6.0-rc.1?

The first thing I noticed is that my normal Text had a padding of 10. And with this new version, there seems to be no padding in that Text...

asturur commented:
https://github.com/kangax/fabric.js/wiki/Changelog-(draft)

updated more or less, maybe latest commits are missing because i have been busy.

Stefanoviic commented:
@asturur thanks.

You know why the iText padding isn't working? See here: https://jsfiddle.net/51u08n39/142/

asturur commented:
Is not working because there i uploaded a too much devel version of fabric.
if you download the current version should work.

Stefanoviic commented:
@asturur ah thanks got both the padding and canvas.toDataURL working now!

I've found a new problem though, which was working fine in 1.5.0.
If you type a very long word without spaces, the boundries of the width of the box are being ignored, see: https://jsfiddle.net/51u08n39/144/

asturur commented:
Yes, of course, because the textbox cannot break words.
There is an ongoing proposal for word breaking:
https://github.com/kangax/fabric.js/issues/2376

with working code. Last version is better.

Stefanoviic commented:
@asturur 

Seems I cant give a fontFamily or fontSize to the whole textbox, only the first character.
Fill and align works properly though. (Worked fine in 1.5.0)

//Set Active Object Name
function setActiveProp(name, value) {
  var object = canvas.getActiveObject();
  if (!object) return;

  object.set(name, value).setCoords();
  canvas.renderAll();
}

//Add Controls
function addControls($scope) {
	
	//Font Family
	$scope.setFontFamily = function(value) {
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));
		$('#fontselect').css("font-family", value);
		setActiveProp('fontFamily', value);
		canvas.renderAll();
	};	
	//Change font color
	$scope.changeFontColor = function() {	  
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));	
		setActiveProp('fill', $('#hexVal').val());
		canvas.renderAll();
	};
	//Align left
  	$scope.alignLeft = function() {
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));	
    	setActiveProp('textAlign', 'left');
  	};
	//Align Center
	$scope.alignCenter = function() {
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));	
    	setActiveProp('textAlign', 'center');
  	};
	//Align Right
	$scope.alignRight = function() {
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));	
    	setActiveProp('textAlign', 'right');
  	};	
	//Font size
	$scope.setFontSize = function(value) {
		var count = canvas.getObjects().length - 1;
  		var object = canvas.setActiveObject(canvas.item(count));
		setActiveProp('fontSize', value);
		canvas.renderAll();
	};		
}

Stefanoviic commented:
Any chance someone knows why only the first character of my textbox is getting the font(size) if selected (See above)?

asturur commented:
there is a bug i noticed and i should verify it again. if you change style of textbox, an array of styles get populated. This will not allow you anymore to change general fontsize.

Please try the latest version from githib, kangax builded it two days ago. then open a separate issue with a jsfiddle demonstrating the problem.

Stefanoviic commented:
@asturur 

Thanks, the newest fabricjs fixed the problem! But an old problem I had is back:

var dataURL = canvas.toDataURL({
format: 'jpeg', 
quality: 1, 
multiplier: 5
});

Seems the multiplier is not working...

Stefanoviic commented:
@asturur you have any idea why the toDateURL is not being multiplied by the multiplier in the latest version?

asturur commented:
not related on this post.

manelio commented:
Is there any news about this feature?

I see the pull request is closed, but I think it's very useful.

asturur commented:
It has been merged and is now on main codebase.

salmanjaved commented:
First of all this is great feature and awesome work. Few question regarding style and word break strategy:
1 - Unable to set style attribute of the object across each character. 
2 - Also the fabricjs 1.6.0 release a breaking strategy on the basis of space. But the breaking strategy must be  based on the number of character and their width like in the demo.

Any reason for these scenarios?

Thanks.

asturur commented:
we implemented a non word breaking strategy. style should work. what
problem do you have with style?
On Apr 18, 2016 7:24 AM, "salmanjaved" <notifications@github.com> wrote:

> First of all this is great feature and awesome work. Few question
> regarding style and word break strategy:
> 1 - Unable to set style attribute of the object across each character.
> 2 - Also the fabricjs 1.6.0 release a breaking strategy on the basis of
> space. But the breaking strategy must be based on the number of character
> and their width like in the demo.
>
> Any reason for these scenarios?
>
> Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kangax/fabric.js/pull/1778#issuecomment-211336653>
>

abhi06991 commented:
Is this feature available in the latest 1.6.2 version, cos it doesnt seem to work

akash31 commented:
How to set fix height and width using fabric.Texbox. 
I am trying to set fix height.

asturur commented:
there is no fixheight.

fahadnabbasi commented:
you need to override fabric.Textbox.prototype._initDimensions function and set the height of Textbox there
`(function() {
//Override function for textbox to handle the height for Textbox class

    fabric.Textbox.prototype._initDimensions = function(ctx) {
        if (this.__skipDimension) {
            return;
        }

        if (!ctx) {
            ctx = fabric.util.createCanvasElement().getContext('2d');
            this._setTextStyles(ctx);
        }

        // clear dynamicMinWidth as it will be different after we re-wrap line
        this.dynamicMinWidth = 0;
        // wrap lines
        this._textLines = this._splitTextIntoLines();
        // if after wrapping, the width is smaller than dynamicMinWidth, change the width and re-wrap
        if (this.dynamicMinWidth > this.width) {
           
                this._set('width', this.dynamicMinWidth);
            
        }
        // clear cache and re-calculate height
        this._clearCache();
        this.minHeight = this._getTextHeight(ctx);
        if(this.height <= 0)
            this.height=this.minHeight;
    }
}).call(this);`

and after that when you are creating the new Textbox object, you need to define the height there

akash31 commented:
How to set **Line wise text align** using fabric.Textbox.
Like., Left, Right, Center.

asturur commented:
you can set it up only for all textbox with the .textAlign property.

www.fabricjs.com/docs

akash31 commented:
Yes I known but **I need text align line index wise.**
any logic ?

asturur commented:
no.

akash31 commented:
Set line height for selected lines.
Is it possible in fabric.Textbox / Text / IText.

astw commented:
@asturur  I tried your link below.  When I try text into textbox,  the textbox size changed if I type fast.  

I noticed your demo is using version 1.6.1.  So I create one using version 1.6.5,  the issue still exists.

Is it a bug?
 
version 1.5.1 http://jsfiddle.net/51u08n39/142/
version 1.6.5 http://jsfiddle.net/astw/bhwypzfo/1/

WilliamCarlos commented:
Is it possible to set a minimum height?

asturur commented:
no. Not a functionality in fabric

WilliamCarlos commented:
D:    Any ideas for a workaround for dynamically adding textboxes to canvas (click and drag)?
Currently I'm setting the width/height of the textbox from mouse:move but setting the height doesn't do much (since it auto-resizes height to the default text).

asturur commented:
You need to extend the object changing some behaviour you do not like. is way better than trying to hack it with events.

Please if you are starting a project, do it on the 2.0 branch, text works way better.

fahadnabbasi commented:
you will have to oveeride init function of textbox n set the height there

On Thu, 15 Jun 2017 at 10:46 PM, William Lee <notifications@github.com>
wrote:

> D: Any ideas for a workaround for dynamically adding textboxes to canvas
> (click and drag)?
> Currently I'm setting the width/height of the textbox from mouse:move but
> setting the height doesn't do much (since it auto-resizes height to the
> default text).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kangax/fabric.js/pull/1778#issuecomment-308790095>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrGntIq00cfYiz0FpoVnF8y0XwcGnpjks5sEVfmgaJpZM4CyIVr>
> .
>
-- 
Fahad Nawaz

IMPMAC commented:
Hi, I am hoping someone can help me here.

I am clicking and then adding a new textbox that is at first empty. I have written code to automatically enter editing mode

My issue is that when you press space, it wraps the text. I only want it to wrap when the user presses return or manually resizes the textbox.

Is there anything I can do/start to get this behaviour?

Thanks
